### ISSUE TITLE
Should warn when calling blanket implementation of a method rather than our own

### ISSUE THREAD

**Issue by gdesmott** [rust-lang/rust#61930]
The following snippet crashes:
```
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
Aborted (core dumped)
```

```rust
use std::fmt;

struct Badger;

impl Badger {
    fn to_string<'a>(self) -> &'a str {
        "badger"
    }
}

impl fmt::Display for Badger {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        f.write_str(&self.to_string())
    }
}

fn main() {
    let b = Badger {};
    println!("{}", b);
}
```

`fmt()` calls the blanket implementation of `ToString`, which is using `Display`, instead of our own implementation resulting in an infinite recursion.

We wouldn't have the problem if our function was taking `&self` but that's still pretty confusing and prone to errors. It would be good to have at least a warning (or a clippy lint?) for such patterns.

**Comment by jonas-schievink** [comment]
This is effectively https://github.com/rust-lang/rust/issues/57965

**Comment by fee1-dead** [comment]
Duplicate of #45838
